cmd/cloud_sql_proxy/internal/healthcheck/healthcheck.go

```diff
@@ -0,0 +1,136 @@
+// limitations under the License.
+
+// Package healthcheck tests and communicates the health of the Cloud SQL Auth proxy.
+package healthcheck
+
+import (
+	"context"
+	"errors"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/logging"
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy"
+)
+
+const (
+	livenessPath = "/liveness"
+	readinessPath = "/readiness"
+)
+
+// Server is a type used to implement health checks for the proxy.
+type Server struct {
+	// startedL protects started.
+	startedL sync.Mutex
+	// started is a flag that indicates whether the proxy is done starting up. 
+	// started is used to support readiness probing and should not be confused 
+	// for affecting startup probing.
+	started bool
+	// port designates the port number on which Server listens and serves.
+	port string
+	// srv is a pointer to the HTTP server used to communicated proxy health.
+	srv *http.Server
+}
+
+// NewServer initializes a Server object and exposes HTTP endpoints used to
+// communicate proxy health.
+func NewServer(c *proxy.Client, port string) (*Server, error) {
+	mux := http.NewServeMux()
+
+	srv := &http.Server{
+		Addr: ":" + port,
+		Handler: mux,
+	}
+
+	hcServer := &Server{
+		port: port,
+		srv:  srv,
+	}
+
+	mux.HandleFunc(readinessPath, func(w http.ResponseWriter, _ *http.Request) {
+		if !isReady(c, hcServer) {
+			w.WriteHeader(500)
+			w.Write([]byte("error"))
+			return
+		}
+		w.WriteHeader(200)
+		w.Write([]byte("ok"))
+	})
+
+	mux.HandleFunc(livenessPath, func(w http.ResponseWriter, _ *http.Request) {
+		if !isLive() { // Because isLive() returns true, this case should not be reached.
+			w.WriteHeader(500)
+			w.Write([]byte("error"))
+			return
+		}
+		w.WriteHeader(200)
+		w.Write([]byte("ok"))
+	})
+
+	ln, err := net.Listen("tcp", srv.Addr)
+	if err != nil {
+		return nil, err
+	}
+	
+	go func() {
+		if err := srv.Serve(ln); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			logging.Errorf("Failed to started health check HTTP server: %v", err)
+		}
+	}()
+
+	return hcServer, nil
+}
+
+// Close gracefully shuts down the HTTP server belonging to the Server object.
+func (s *Server) Close(ctx context.Context) error {
+	return s.srv.Shutdown(ctx)
+}
+
+// NotifyStarted tells the Server that the proxy has finished startup.
+func (s *Server) NotifyStarted() {
+	s.startedL.Lock()
+	s.started = true
+	s.startedL.Unlock()
+}
+
+// isLive returns true as long as the proxy is running.
+func isLive() bool {
+	return true
+}
+
+// isReady will check the following criteria before determining whether the
+// proxy is ready for new connections.
+// 1. Finished starting up / been sent the 'Ready for Connections' log.
+// 2. Not yet hit the MaxConnections limit, if applicable.
+func isReady(c *proxy.Client, s *Server) bool {
+	// Not ready until we reach the 'Ready for Connections' log.
+	s.startedL.Lock()
+	started := s.started
+	s.startedL.Unlock()
```

Comment:
nit: I think we have an opportunity to use a channel here (instead of a mutex) so that we aren't blocking. 

Right now, every called to `s.startedL` means that thread blocks until it has access to the mutex. However in this case that's pretty wasteful, because we only write to the value once, and read it after that. 

If we use a signal channel, we can just close the channel when we are "ready":
```go
close(s.started)
```

And use a select statement to determine if started is `true` (closed) or `false`(open):
```go
select {
    case <-s.started:
        // only returns when closed, so this means the proxy has started
        return true
     default:
         // only hits default when open, so false
}
return false
```

We can tell if the proxy is ready by if the channel is open or not.